libraries/SITL/SIM_GPS.cpp

```diff
@@ -998,198 +996,447 @@ uint32_t GPS_NOVA::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_
+                GSOF_VEL_LEN,
+                vel_flags,
+                gsof_pack_float(d->speed_2d()),
+                gsof_pack_float(d->heading()),
+                // Trimble API has ambiguous direction here.
+                // Intentionally narrow from double.
+                gsof_pack_float(static_cast<float>(d->speedD))
+            };
+            static_assert(sizeof(gsof_vel) - (sizeof(gsof_vel::OUTPUT_RECORD_TYPE) + sizeof(gsof_vel::RECORD_LEN)) == GSOF_VEL_LEN);
+
+            payload_sz += sizeof(vel);
+            memcpy(&buf[offset], &vel, sizeof(vel));
+            offset += sizeof(vel);
+        }
+    }
+    if (channel_rates[uint8_t(Gsof_Msg_Record_Type::PDOP_INFO)] != Output_Rate::OFF){
+        const auto last_time = last_publish_ms[uint8_t(Gsof_Msg_Record_Type::PDOP_INFO)];
+        const auto desired_rate = channel_rates[uint8_t(Gsof_Msg_Record_Type::PDOP_INFO)];
+
+        if (now - last_time > RateToPeriodMs(desired_rate)) {
+            // https://receiverhelp.trimble.com/oem-gnss/index.html#GSOFmessages_PDOP.html?TocPath=Output%2520Messages%257CGSOF%2520Messages%257C_____12
+            constexpr uint8_t GSOF_DOP_LEN = 0x10;
+            const struct PACKED gsof_dop {
+                const uint8_t OUTPUT_RECORD_TYPE { uint8_t(Gsof_Msg_Record_Type::PDOP_INFO) }; 
+                const uint8_t RECORD_LEN { GSOF_DOP_LEN };
+                uint32_t pdop = htobe32(1);
+                uint32_t hdop = htobe32(1);
+                uint32_t vdop = htobe32(1);
+                uint32_t tdop = htobe32(1);
+            } dop {};
+            // Check the payload size calculation in the compiler
+            constexpr auto dop_size = sizeof(gsof_dop);
+            static_assert(dop_size == 18);
+            constexpr auto dop_record_type_size = sizeof(gsof_dop::OUTPUT_RECORD_TYPE);
+            static_assert(dop_record_type_size == 1);
+            constexpr auto len_size = sizeof(gsof_dop::RECORD_LEN);
+            static_assert(len_size == 1);
+            constexpr auto dop_payload_size = dop_size - (dop_record_type_size + len_size);
+            static_assert(dop_payload_size == GSOF_DOP_LEN);
+
+            payload_sz += sizeof(dop);
+            memcpy(&buf[offset], &dop, sizeof(dop));
+            offset += sizeof(dop);
+        }
+    }
+    if (channel_rates[uint8_t(Gsof_Msg_Record_Type::POSITION_SIGMA_INFO)] != Output_Rate::OFF){
+        const auto last_time = last_publish_ms[uint8_t(Gsof_Msg_Record_Type::POSITION_SIGMA_INFO)];
+        const auto desired_rate = channel_rates[uint8_t(Gsof_Msg_Record_Type::POSITION_SIGMA_INFO)];
+        if (now - last_time > RateToPeriodMs(desired_rate)) {
+            // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_SIGMA.html
+            constexpr uint8_t GSOF_POS_SIGMA_LEN = 0x26;
+            const struct PACKED gsof_pos_sigma {
+                const uint8_t OUTPUT_RECORD_TYPE { uint8_t(Gsof_Msg_Record_Type::POSITION_SIGMA_INFO) }; 
```

Comment:
I think the cast to uint8_t is probably unnecessary here (although it does no harm).